cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,93 @@
+# cuML developer guide
+This document summarizes rules and best practices for contributions to C++ component cuML of RAPIDS/cuml.
+
+# General
+
+# Thread safety
+
+# Coding style
+
+# Error handling
+All calls to CUDA APIs should be done via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. Those macros take care of checking the return values of the used API calls and generate an exception in case the command was no successful. In case an exception needs to be avoided, e.g. when implementing a destructor, `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVER_CHECK_NO_THROW ` (currently not available, see https://github.com/rapidsai/cuml/issues/229) should be used. Those macros will only log the error but do not throw an exception.
+
+# Logging
+Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+
+# Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->setDeviceAllocator()->allocate(n*sizeof(T), stream);
```

Comment:
I believe you meant `getDeviceAllocator`?

Reply:
Yes I fixed it.